### PR TITLE
feat: implement kubernetes node audit controller

### PR DESCRIPTION
**Pull request by utkuozdemir**
Clean up invalid nodes (from the Kubernetes perspective) from the cluster if they are no longer part of a cluster from the perspective of Omni.

Introduce a new intermediate resource, `ClusterKubernetesNodes`, containing the ordered list of nodes in a cluster.

Create the resource, `KubernetesNodeAuditResult`, to provide a summary on the latest deleted nodes from the Kubernetes cluster.

Closes siderolabs/omni#9.

### CODE REVIEW

File: internal/backend/runtime/omni/controllers/omni/cluster_kubernetes_nodes.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+					return nil
+				}
+
+				nodes := make([]string, 0, identityList.Len())
+
+				for iter := identityList.Iterator(); iter.Next(); {
+					identity := iter.Value()
+
+					if identity.TypedSpec().Value.Nodename != "" {
```
Comment (Unix4ever) [inline internal/backend/runtime/omni/controllers/omni/cluster_kubernetes_nodes.go:59]: I wonder if we should skip audit if this controller finds the nodes without a nodename.
Normally it should never be empty and if it is we might want to hold off the audit until it gets fixed.

File: internal/backend/runtime/omni/controllers/omni/omni_test.go [hunk 2]
```diff
@@ -347,7 +347,7 @@ func (suite *OmniSuite) newServer(suffix string, opts ...grpc.ServerOption) (*ma
 }
 
 func (suite *OmniSuite) SetupTest() {
-	suite.ctx, suite.ctxCancel = context.WithTimeout(context.Background(), 20*time.Second)
+	suite.ctx, suite.ctxCancel = context.WithTimeout(context.Background(), 20*time.Hour) // todo: temp, revert
```
Comment (Unix4ever) [inline internal/backend/runtime/omni/controllers/omni/omni_test.go:350]: Don't forget to revert :)

**Review by smira — APPROVED** [review]
I think we should have an integration test for it.

Also we need to skip deleting nodes which are Ready?

**Review by Unix4ever — APPROVED** [review]

File: internal/backend/runtime/omni/controllers/omni/omni_test.go (hunk 2 above)
Comment (utkuozdemir) [reply internal/backend/runtime/omni/controllers/omni/omni_test.go:350]: good point :D

**Comment by utkuozdemir** [comment]
> I think we should have an integration test for it.
> 
> Also we need to skip deleting nodes which are Ready?

already doing the `Ready` check here:
https://github.com/siderolabs/omni/pull/150/files#diff-f856c47e45cecba4018f26b99f4529186e842968b8a8a6acf6201f6bbaec9d61R230-R234

Will add an integration test.

**Comment by Unix4ever** [comment]
Just remembered: we also need to drop `MachineSet` dependency in the cluster machine teardown controller.
But please run all e2e tests after that: especially e2e scaling.

**Comment by utkuozdemir** [comment]
> Just remembered: we also need to drop `MachineSet` dependency in the cluster machine teardown controller. But please run all e2e tests after that: especially e2e scaling.

I see, but why? - We still remove the node in that controller as well. This audit controller is just additional.

File: internal/backend/runtime/omni/controllers/omni/cluster_kubernetes_nodes.go (hunk 1 above)
Comment (utkuozdemir) [reply internal/backend/runtime/omni/controllers/omni/cluster_kubernetes_nodes.go]: done - we don't skip empties here anymore, and the other controller skips reconciliation if there's an empty one.

**Pull request merged**